### PR TITLE
Add Console plugin

### PR DESCRIPTION
**Pull request by Scarface1809**
# Description

Create a console plugin to be able to visualize the logs in tesseratos.

### CODE REVIEW

File: tools/tesseratos/include/tesseratos/console/plugin.hpp [hunk 1]
```diff
@@ -0,0 +1,26 @@
...
+/// @brief Plugin entry point.
+/// @ingroup tesseratos-console-plugin
+
+#pragma once
+
+#include <cubos/engine/assets/plugin.hpp>
+
+namespace tesseratos
+{
+    /// @defgroup tesseratos-console-plugin Asset explorer
```
Comment (DiogoMendonc-a) [inline tools/tesseratos/include/tesseratos/console/plugin.hpp:14]: This is not the asset explorer

File: tools/tesseratos/src/tesseratos/console/plugin.cpp [hunk 2]
```diff
@@ -0,0 +1,290 @@
...
+    case cubos::core::Logger::Level::Debug:
+        return ImVec4(0.0F, 1.0F, 1.0F, 1.0F); // Cyan
+    case cubos::core::Logger::Level::Info:
+        return ImVec4(0.0F, 1.0F, 0.0F, 1.0F); // Green
+    case cubos::core::Logger::Level::Warn:
+        return ImVec4(1.0F, 1.0F, 0.0F, 1.0F); // Yellow
+    case cubos::core::Logger::Level::Error:
+        return ImVec4(1.0F, 0.0F, 0.0F, 1.0F); // Red
+    case cubos::core::Logger::Level::Critical:
+        return ImVec4(1.0F, 1.0F, 1.0F, 1.0F); // White (or any color you prefer for critical)
```
Comment (DiogoMendonc-a) [inline tools/tesseratos/src/tesseratos/console/plugin.cpp:66]: Shouldn't critical be red?

File: tools/tesseratos/include/tesseratos/console/plugin.hpp [hunk 3]
```diff
@@ -0,0 +1,26 @@
...
+    /// @ingroup tesseratos
+    /// @brief Displays Cubos Logs.
+    ///
+    /// ## Dependencies
+    /// - @ref imgui-plugin
+    /// - @ref log-plugin
```
Comment (DiogoMendonc-a) [inline tools/tesseratos/include/tesseratos/console/plugin.hpp:20]: This is not a plugin, right?

File: tools/tesseratos/src/tesseratos/console/plugin.cpp [hunk 4]
```diff
@@ -0,0 +1,290 @@
...
+        ImGui::BeginChild("Log", ImVec2(0, -ImGui::GetFrameHeightWithSpacing()));
+
+        while (cubos::core::Logger::read(cursor, entry))
+        {
+            ui_entries.push_back(entry);
+        }
+
+        if (mode.clearEnabled)
+        {
+            ui_entries.clear();
```
Comment (DiogoMendonc-a) [inline tools/tesseratos/src/tesseratos/console/plugin.cpp:117]: You seem to be resetting this vector each fram anyway, wouldn't it be better to just not push the entries in the first place when the console is to be cleaned?

File: tools/tesseratos/src/tesseratos/console/plugin.cpp [hunk 5]
```diff
@@ -0,0 +1,290 @@
...
+        // Log Scrollable Window
+        ImGui::SetColumnWidth(0, ImGui::GetWindowWidth() * 0.8f);
+
+        std::size_t cursor = mode.prevCursor;
+        cubos::core::Logger::Entry entry;
+        std::vector<cubos::core::Logger::Entry> ui_entries;
+        cubos::core::Logger::Entry previousEntry;
+
+        // Counters for each log level
+        std::size_t traceCount = 0;
```
Comment (DiogoMendonc-a) [inline tools/tesseratos/src/tesseratos/console/plugin.cpp:101]: Wouldn't it be better to use a map here intead of hardcoded variables?

**Review by DiogoMendonc-a — CHANGES_REQUESTED** [review]

**Comment by roby2014** [comment]
Please add context to PR description. As in, what is the main goal of this, what was the strategy/plan applied, why is it needed..

File: tools/tesseratos/src/tesseratos/console/plugin.cpp [hunk 6]
```diff
@@ -0,0 +1,290 @@
...
+        ImGui::Begin("Console");
+
+        ImGui::Columns(2, "ConsoleLayout", true);
+
+        // Log Scrollable Window
+        ImGui::SetColumnWidth(0, ImGui::GetWindowWidth() * 0.8f);
+
+        std::size_t cursor = mode.prevCursor;
+        cubos::core::Logger::Entry entry;
+        std::vector<cubos::core::Logger::Entry> ui_entries;
```
Comment (RiscadoA) [inline tools/tesseratos/src/tesseratos/console/plugin.cpp:97]: This vector should be stored in a resource (Mode?), no need to recreate it every frame.

File: tools/tesseratos/src/tesseratos/console/plugin.cpp [hunk 7]
```diff
@@ -0,0 +1,290 @@
...
+#include <tesseratos/console/plugin.hpp>
+#include <tesseratos/toolbox/plugin.hpp>
+
+using cubos::engine::Cubos;
+
+using namespace tesseratos;
+
+namespace
+{
+    struct Mode
```
Comment (RiscadoA) [inline tools/tesseratos/src/tesseratos/console/plugin.cpp:16]:
```suggestion
    struct State
```
for consistency with other plugins

**Review by RiscadoA — CHANGES_REQUESTED** [review]

File: tools/tesseratos/include/tesseratos/console/plugin.hpp (hunk 3 above)
Comment (Scarface1809) [reply tools/tesseratos/include/tesseratos/console/plugin.hpp:20]: Todos os outros têm commented isso so im not sure mas diria que sim

File: tools/tesseratos/src/tesseratos/console/plugin.cpp [hunk 8]
```diff
@@ -0,0 +1,291 @@
...
+#include <tesseratos/console/plugin.hpp>
+#include <tesseratos/toolbox/plugin.hpp>
+
+using cubos::engine::Cubos;
+
+using namespace tesseratos;
+
+namespace
+{
+    struct State
```
Comment (roby2014) [inline tools/tesseratos/src/tesseratos/console/plugin.cpp:16]: add documentation

File: tools/tesseratos/src/tesseratos/console/plugin.cpp [hunk 9]
```diff
@@ -0,0 +1,291 @@
...
+        bool traceEnabled = true;
+        bool debugEnabled = true;
+        bool infoEnabled = true;
+        bool warnEnabled = true;
+        bool errorEnabled = true;
+        bool criticalEnabled = true;
+    };
+} // namespace
+
+static std::string levelToString(cubos::core::Logger::Level level)
```
Comment (roby2014) [inline tools/tesseratos/src/tesseratos/console/plugin.cpp:31]: add documentation

File: tools/tesseratos/src/tesseratos/console/plugin.cpp [hunk 10]
```diff
@@ -0,0 +1,291 @@
...
+    case cubos::core::Logger::Level::Error:
+        return "Error";
+    case cubos::core::Logger::Level::Critical:
+        return "Critical";
+    default:
+        return "Unknown";
+    }
+}
+
+static ImVec4 levelToColor(cubos::core::Logger::Level level)
```
Comment (roby2014) [inline tools/tesseratos/src/tesseratos/console/plugin.cpp:52]: add documentation

File: tools/tesseratos/src/tesseratos/console/plugin.cpp [hunk 11]
```diff
@@ -0,0 +1,291 @@
...
+    cubos.depends(toolboxPlugin);
+    cubos.resource<State>();
+
+    cubos.system("show Logger UI").tagged(cubos::engine::imguiTag).call([](Toolbox& toolbox, State& state) {
+        if (!toolbox.isOpen("Console"))
+        {
+            return;
+        }
+
+        static char searchString[256];
```
Comment (roby2014) [inline tools/tesseratos/src/tesseratos/console/plugin.cpp:85]: why also not store it in state?

File: tools/tesseratos/src/tesseratos/console/plugin.cpp [hunk 12]
```diff
@@ -0,0 +1,291 @@
...
+            case cubos::core::Logger::Level::Error:
+                errorCount++;
+                break;
+            case cubos::core::Logger::Level::Critical:
+                criticalCount++;
+                break;
+            case cubos::core::Logger::Level::Off:
+                break;
+            }
+
```
Comment (roby2014) [inline tools/tesseratos/src/tesseratos/console/plugin.cpp:150]: cant this be simplified or written diferently? e.g using an unordered map:

```cpp
std::unordered_map<cubos::core::Logger::Level, int> logCounts = {
    {cubos::core::Logger::Level::Trace, 0},
    {cubos::core::Logger::Level::Debug, 0},
    {cubos::core::Logger::Level::Info, 0},
    {cubos::core::Logger::Level::Warn, 0},
    {cubos::core::Logger::Level::Error, 0},
    {cubos::core::Logger::Level::Critical, 0}
};

for (cubos::core::Logger::Entry& e : state.ui_entries)
{
    if (e.level != cubos::core::Logger::Level::Off)
    {
        logCounts[e.level]++;
    }
}
```

**Review by roby2014 — CHANGES_REQUESTED** [review]
sorry for the delay. review done, some nitpicks and i feel like there is a bit of repetitive code which can be designed differently, but open for further discussion :)

**Comment by Scarface1809** [comment]
I dont get why it fails in the lint :cold_sweat:

**Comment by RiscadoA** [comment]
> I dont get why it fails in the lint 😰

Disregard it, that action is very buggy :/ fixed it in another PR, but it wasnt merged yet

File: tools/tesseratos/src/tesseratos/console/plugin.cpp [hunk 13]
```diff
@@ -0,0 +1,270 @@
...
+        while (cubos::core::Logger::read(cursor, entry))
+        {
+            state.uiEntries.push_back(entry);
+        }
+
+        if (state.clearEnabled)
+        {
+            state.clearEnabled = false;
+            state.prevCursor = cursor;
+        }
```
Comment (RiscadoA) [inline tools/tesseratos/src/tesseratos/console/plugin.cpp:122]:
```suggestion
        cubos::core::Logger::Entry entry;
        cubos::core::Logger::Entry previousEntry;

        // Counters for each log level
        std::unordered_map<cubos::core::Logger::Level, int> logCounts = {
            {cubos::core::Logger::Level::Trace, 0}, {cubos::core::Logger::Level::Debug, 0},
            {cubos::core::Logger::Level::Info, 0},  {cubos::core::Logger::Level::Warn, 0},
            {cubos::core::Logger::Level::Error, 0}, {cubos::core::Logger::Level::Critical, 0}};

        ImGui::BeginChild("Log", ImVec2(0, -ImGui::GetFrameHeightWithSpacing()));

        while (cubos::core::Logger::read(state.prevCursor, entry))
        {
            state.uiEntries.push_back(entry);
        }

        if (state.clearEnabled)
        {
            state.clearEnabled = false;
            state.uiEntries.clear();
        }
```

You want to store the `cursor` written to by `Logger::read`. This way, on each call to this system (on each frame), you only push_back new log entries to `uiEntries`, and thus you don't need to clear it every frame! This is way more efficient.
When you want to forget 'past history', you simply clear the `uiEntries` vector.

File: tools/tesseratos/src/tesseratos/console/plugin.cpp [hunk 14]
```diff
@@ -0,0 +1,270 @@
...
+
+        // Collapse Button
+        if (ImGui::SmallButton("Collapse"))
+        {
+            state.collapseEnabled = !state.collapseEnabled;
+        }
+
+        // Log Level Counters
+        // ImGui SELECTABLE
```
Comment (RiscadoA) [inline tools/tesseratos/src/tesseratos/console/plugin.cpp:230]: ?

File: tools/tesseratos/src/tesseratos/console/plugin.cpp [hunk 15]
```diff
@@ -0,0 +1,270 @@
...
+
+using namespace tesseratos;
+
+namespace
+{
+    // Resource used to store the state of the console
+    struct State
+    {
+        std::vector<cubos::core::Logger::Entry> uiEntries;
+        std::size_t prevCursor = 0; // Previous cursor position~could find a better way to clear log
```
Comment (RiscadoA) [inline tools/tesseratos/src/tesseratos/console/plugin.cpp:21]: Considering my comment below, a better name would be just `cursor`, I think.

**Review by RiscadoA — CHANGES_REQUESTED** [review]

File: tools/tesseratos/src/tesseratos/console/plugin.cpp [hunk 16]
```diff
@@ -0,0 +1,270 @@
...
+    case cubos::core::Logger::Level::Debug:
+        return {0.0F, 1.0F, 1.0F, 1.0F}; // Cyan
+    case cubos::core::Logger::Level::Info:
+        return {0.0F, 1.0F, 0.0F, 1.0F}; // Green
+    case cubos::core::Logger::Level::Warn:
+        return {1.0F, 1.0F, 0.0F, 1.0F}; // Yellow
+    case cubos::core::Logger::Level::Error:
+        return {1.0F, 0.0F, 0.0F, 1.0F}; // Red
+    case cubos::core::Logger::Level::Critical:
+        return {1.0F, 0.0F, 0.0F, 0.5F}; // Bright Red
```
Comment (DiogoMendonc-a) [inline tools/tesseratos/src/tesseratos/console/plugin.cpp:72]: How is this Bright Red? Is it not just the same red as Error, but more transparent?

File: tools/tesseratos/src/tesseratos/console/plugin.cpp [hunk 17]
```diff
@@ -0,0 +1,270 @@
...
+    default:
+        return {1.0F, 1.0F, 1.0F, 1.0F}; // Default to
```
Comment (DiogoMendonc-a) [inline tools/tesseratos/src/tesseratos/console/plugin.cpp:74]:
```suggestion
        return {1.0F, 1.0F, 1.0F, 1.0F}; // Default to white
```

**Review by DiogoMendonc-a — APPROVED** [review]
Nothing much to add to what was already said, just a couple of small details.

**Review by RiscadoA — CHANGES_REQUESTED** [review]
LGTM! It's just missing an entry in the CHANGELOG.md, will merge it afterwards.

**Review by roby2014 — APPROVED** [review]

**Review by RiscadoA — APPROVED** [review]

**Pull request merged**